mesonbuild/interpreter.py

```diff
@@ -1545,6 +1547,24 @@ def first_supported_link_argument_method(self, args, kwargs):
         mlog.log('First supported link argument:', mlog.red('None'))
         return []
 
+    @FeatureNew('compiler.has_function_attribute', '0.48.0')
+    @permittedKwargs({})
+    def has_func_attribute_method(self, args, kwargs):
+        args = mesonlib.stringlistify(args)
+        if len(args) != 1:
+            raise InterpreterException('has_func_attribute takes exactly one argument.')
+        result = self.compiler.has_func_attribute(args[0], self.environment)
+        h = mlog.green('YES') if result else mlog.red('NO')
+        mlog.log('Compiler for {} supports function attribute {}:'.format(self.compiler.get_display_language(), args[0]), h)
+        return result
+
+    @FeatureNew('compiler.get_supported_function_attributes', '0.48.0')
+    @permittedKwargs({})
+    def get_supported_function_attributes_method(self, args, kwargs):
+        args = mesonlib.stringlistify(args)
+        return [a for a in args if self.has_func_attribute_method(a, kwargs)]
```

Comment:
Just an aside: we could improve the performance of this in the future by checking many of the attributes at once (and caching the positive matches).

Reply:
I think it would be worthwhile to cache these as well since (other than ifunc) they're pure compiler features, and it would speed up rebuilds.

Reply:
Sure, I meant that if we can check many attributes at once, we can also cache all of them at once.